base/src/main/java/org/mozilla/jss/pkcs12/MacData.java

```diff
@@ -115,18 +141,31 @@ public MacData( Password password, byte[] macSalt,
 
         try {
             // generate key from password and salt
-            KeyGenerator kg = token.getKeyGenerator(KeyGenAlgorithm.PBA_SHA1_HMAC);
+            if(algID == null) {
+                algID = new AlgorithmIdentifier(DigestAlgorithm.SHA1.toOID());
+            }
+            KeyGenerator kg = null;
+            JSSMessageDigest digest = null;
+            if(DigestAlgorithm.SHA1.toOID().equals(algID.getOID())){
+                kg = token.getKeyGenerator(KeyGenAlgorithm.PBA_SHA1_HMAC);
+                digest = token.getDigestContext(HMACAlgorithm.SHA1);
+            }
+            if(DigestAlgorithm.SHA256.toOID().equals(algID.getOID())){
+                kg = token.getKeyGenerator(KeyGenAlgorithm.PBE_SHA256_HMAC);
+                digest = token.getDigestContext(HMACAlgorithm.SHA256);
+            }
+            if(kg == null) {
```

Comment:
Is it possible for the digest to be null as well?

Reply:
Actually, both `kg` and `digest` should be null or both should have a value because if one of them is not possible than an exception is raised and this code is not reached. So for the check one of them  should be enough but  I am going to include both in the `if` for more safety.

Reply:
Looks good.